metadata/src/main/java/org/apache/kafka/controller/DelegationTokenControlManager.java

```diff
@@ -0,0 +1,359 @@
+        }
+
+        long renewLifeTime = tokenDefaultRenewLifetime;
+        if (requestData.renewPeriodMs() > 0) {
+            renewLifeTime = Math.min(renewLifeTime, requestData.renewPeriodMs());
+        }
+        long renewTimeStamp = now + renewLifeTime;
+        long expiryTimestamp = Math.min(myTokenInformation.maxTimestamp(), renewTimeStamp);
+
+        myTokenInformation.setExpiryTimestamp(expiryTimestamp);
+
+        DelegationTokenData newDelegationTokenData = new DelegationTokenData(myTokenInformation);
+
+        responseData
+            .setErrorCode(NONE.code())
+            .setExpiryTimestampMs(expiryTimestamp);
+
+        records.add(new ApiMessageAndVersion(newDelegationTokenData.toRecord(), (short) 0));
+        return ControllerResult.atomicOf(records, responseData);
+    }
+
+    public ControllerResult<ExpireDelegationTokenResponseData> expireDelegationToken(
+        ControllerRequestContext context,
+        ExpireDelegationTokenRequestData requestData,
+        MetadataVersion metadataVersion
+    ) {
+        long now = time.milliseconds();
+
+        List<ApiMessageAndVersion> records = new ArrayList<>();
+        ExpireDelegationTokenResponseData responseData = new ExpireDelegationTokenResponseData();
+
+        if (secretKeyString == null) {
+            // DelegationTokens are not enabled
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_AUTH_DISABLED.code()));
+        }
+
+        TokenInformation myTokenInformation = getToken(requestData.hmac());
+
+        if (myTokenInformation == null) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_NOT_FOUND.code()));
+        }
+
+        if (myTokenInformation.maxTimestamp() < now || myTokenInformation.expiryTimestamp() < now) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_EXPIRED.code()));
+        }
+
+        if (!allowedToRenew(myTokenInformation, context.principal())) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_OWNER_MISMATCH.code()));
+        }
+
+        if (requestData.expiryTimePeriodMs() < 0) { // expire immediately
+            responseData
+                .setErrorCode(NONE.code())
+                .setExpiryTimestampMs(requestData.expiryTimePeriodMs());
+            records.add(new ApiMessageAndVersion(new RemoveDelegationTokenRecord().
+                setTokenId(myTokenInformation.tokenId()), (short) 0));
+        } else {
+            long expiryTimestamp = Math.min(myTokenInformation.maxTimestamp(),
+                now + requestData.expiryTimePeriodMs());
+
+            responseData
+                .setErrorCode(NONE.code())
+                .setExpiryTimestampMs(expiryTimestamp);
+
+            myTokenInformation.setExpiryTimestamp(expiryTimestamp);
```

Comment:
Same question as above -- should this really be done here, or should it be done when the record gets replayed?

Reply:
It does get updated on replay. Is `(myTokenInformation` a reference or a copy and how can I tell. I want a copy.

Reply:
Fixed